cmd/gcp-controller-manager/service_account_verifier.go

```diff
@@ -0,0 +1,414 @@
+// (GSA) and persisting the authorized pairs in a ConfigMap entry.
+//
+// See go/gke-direct-path-controller-dd for details.
+//
+// verifiedSAs is a thread-safe map of all authorized {KSA: GSA} pairs.  This control loop also
+// maintains these pairs in ConfigMap "verifiedSAConfigMapNamespace/verifiedSAConfigMapName".
+type serviceAccountVerifier struct {
+	c           clientset.Interface
+	sals        corelisters.ServiceAccountLister
+	cmls        corelisters.ConfigMapLister
+	saHasSynced func() bool
+	saQueue     workqueue.RateLimitingInterface
+	cmQueue     workqueue.RateLimitingInterface
+	verifiedSAs *saMap
+	hms         *hmsClient
+}
+
+func newServiceAccountVerifier(client clientset.Interface, saInformer coreinformers.ServiceAccountInformer, cmInformer coreinformers.ConfigMapInformer, cs *compute.Service, sm *saMap, hmsAuthzURL string) (*serviceAccountVerifier, error) {
+	hms, err := newHMSClient(hmsAuthzURL, &clientcmdapi.AuthProviderConfig{Name: "gcp"})
+	if err != nil {
+		return nil, err
+	}
+	sav := &serviceAccountVerifier{
+		c:           client,
+		sals:        saInformer.Lister(),
+		cmls:        cmInformer.Lister(),
+		saHasSynced: saInformer.Informer().HasSynced,
+		saQueue: workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 1000*time.Second),
+		), saVerifierSAQueueName),
+		cmQueue: workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 1000*time.Second),
+		), saVerifierCMQueueName),
+		verifiedSAs: sm,
+		hms:         hms,
+	}
+	saInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    sav.onSAAdd,
+		UpdateFunc: sav.onSAUpdate,
+		DeleteFunc: sav.onSADelete,
+	}, serviceAccountResyncPeriod)
+	cmInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    sav.onCMAdd,
+		UpdateFunc: sav.onCMUpdate,
+		DeleteFunc: sav.onCMDelete,
+	}, configMapResyncPeriod)
+	return sav, nil
+}
+
+func (sav *serviceAccountVerifier) onSAAdd(obj interface{}) {
+	klog.V(5).Infof("onSAAdd: %v", obj)
+	sav.enqueueSA(obj)
+}
+
+func (sav *serviceAccountVerifier) onSAUpdate(obj, oldObj interface{}) {
+	klog.V(5).Infof("onSAUpdate: %v -> %v", oldObj, obj)
+	// Enqueues for verification even if no change because of periodic resync.
+	// TODO(danielywong): confirm this is called upon resync
+	sav.enqueueSA(obj)
+}
+
+func (sav *serviceAccountVerifier) onSADelete(obj interface{}) {
+	klog.V(5).Infof("onSADelete: %v", obj)
+	// TODO(danielywong): check delete behavior; expect obj still exist in cache store
+	sav.enqueueSA(obj)
+}
+
+func (sav *serviceAccountVerifier) enqueueSA(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("internal error. Couldn't get key for ServiceAccount %+v: %v", obj, err))
```

Comment:
You can just `klog.Errorf`, this utilruntime package doesn't add much value in our case.
Same below.